misinformation/multimodal_search.py

```diff
@@ -322,7 +348,293 @@ def multimodal_search(
+            cams = (
+                cams[:, :, :, 1:].reshape(visual_input.size(0), 12, -1, 24, 24) * mask
+            )
+            grads = (
+                grads[:, :, :, 1:]
+                .clamp(0)
+                .reshape(visual_input.size(0), 12, -1, 24, 24)
+                * mask
+            )
+
+            gradcam = cams * grads
+            # [enc token gradcam, average gradcam across token, gradcam for individual token]
+            # gradcam = torch.cat((gradcam[0:1,:], gradcam[1:token_length+1, :].sum(dim=0, keepdim=True)/token_length, gradcam[1:, :]))
+            gradcam = gradcam.mean(1).cpu().detach()
+            gradcam = (
+                gradcam[:, 1 : token_length + 1, :].sum(dim=1, keepdim=True)
+                / token_length
+            )
+
+        return gradcam, output
+
+    def resize_img(self, raw_img):
+        w, h = raw_img.size
+        scaling_factor = 240 / w
+        resized_image = raw_img.resize(
+            (int(w * scaling_factor), int(h * scaling_factor))
+        )
+        return resized_image
+
+    def getAttMap(self, img, attMap, blur=True, overlap=True):
+        attMap -= attMap.min()
+        if attMap.max() > 0:
+            attMap /= attMap.max()
+        attMap = skimage_transform.resize(
+            attMap, (img.shape[:2]), order=3, mode="constant"
+        )
+        if blur:
+            attMap = filters.gaussian_filter(attMap, 0.02 * max(img.shape[:2]))
+            attMap -= attMap.min()
+            attMap /= attMap.max()
+        cmap = plt.get_cmap("jet")
+        attMapV = cmap(attMap)
+        attMapV = np.delete(attMapV, 3, 2)
+        if overlap:
+            attMap = (
+                1 * (1 - attMap**0.7).reshape(attMap.shape + (1,)) * img
+                + (attMap**0.7).reshape(attMap.shape + (1,)) * attMapV
+            )
+        return attMap
+
+    def upload_model_blip2_coco(self):
+
+        itm_model = load_model(
+            "blip2_image_text_matching",
+            "coco",
+            is_eval=True,
+            device=MultimodalSearch.multimodal_device,
+        )
+        vis_processor = load_processor("blip_image_eval").build(image_size=364)
+        return itm_model, vis_processor
+
+    def upload_model_blip_base(self):
+
+        itm_model = load_model(
+            "blip_image_text_matching",
+            "base",
+            is_eval=True,
+            device=MultimodalSearch.multimodal_device,
+        )
+        vis_processor = load_processor("blip_image_eval").build(image_size=384)
+        return itm_model, vis_processor
+
+    def upload_model_blip_large(self):
+
+        itm_model = load_model(
+            "blip_image_text_matching",
+            "large",
+            is_eval=True,
+            device=MultimodalSearch.multimodal_device,
+        )
+        vis_processor = load_processor("blip_image_eval").build(image_size=384)
+        return itm_model, vis_processor
+
+    #    @profile
+    def image_text_match_reordering(
```

Comment:
is it possible to simplify this by using helper methods, then we can also add tests for part of this?